### PR TITLE
uCIFI Data Model 2.0: Adding new objects

### PR DESCRIPTION
**Pull request by gmazzoleni**
**New objects**

Connectivity
	3447 - LPWAN Mesh Connectivity
	3448 - LPWAN Mesh Statistics

Sensors
	3422 - Three-phase electrical meter complement
	3445 - Stream Gauge
	3446 - Rain Gauge
	3449 - Anemometer
	3450 - Wet Bulb Globe Temperature
	3451 - Roll over sensor

Schedules
	3452 - Program Scheduler
	3453 - Calendar Rule
	3454 - Program Function
	3455 - Input Filter
	3456 - Output Filter
	3457 - Program Manager
	3458 - Program Monitor

Device to Device communication
	3459 - Distributed Sensor Group
	3460 - Distributed Sensor Message
	3461 - Sensor Group Device Capabilities

### CODE REVIEW

File: 3422.xml [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+				<RangeEnumeration/>
+				<Units>VAh</Units>
+				<Description><![CDATA[Apparent energy export, unified rate, on the 3 lines.]]></Description>
+			</Item>
+			<Item ID="29">
+				<Name>Number of power failures in any phase</Name>
+				<Operations>R</Operations>
+				<MultipleInstances>Single</MultipleInstances>
+				<Mandatory>Optional</Mandatory>
+				<Type>Float</Type>
```
Comment (akeranen) [inline 3422.xml]: Can the number of failures be a non-integer value? Should this be an unsigned integer instead of float?

File: 3452.xml [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+				<Type>Integer</Type>
+				<Operations>RW</Operations>
+				<Mandatory>Mandatory</Mandatory>
+				<MultipleInstances>Single</MultipleInstances>
+				<Description><![CDATA[Number of Program Function object instance containing information on how to control the output(s) based on time and sensor input(s)]]></Description>
+				<Units/>
+			</Item>
+			<Item ID="5">
+				<Name>Input Sources</Name>
+				<Type>String</Type>
```
Comment (akeranen) [inline 3452.xml:98]: Would CoRE link make sense here as type?

File: 3459.xml [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		<MultipleInstances>Multiple</MultipleInstances>
+		<Mandatory>Optional</Mandatory>
+		<Resources>
+			<Item ID="0">
+				<Name>Group State</Name>
+				<Type>Float</Type>
+				<Operations>R</Operations>
+				<Mandatory>Optional</Mandatory>
+				<MultipleInstances>Single</MultipleInstances>
+				<Description><![CDATA[The current output state of the sensor Group, derived either from the latest sensor state published to the group, or computed based on all sensor inputs when data Fusion Operation is used [default=NaN]]]></Description>
```
Comment (akeranen) [inline 3459.xml:50]: How is NaN represented on the wire?
(same comment applies to other objects too)

File: 3459.xml [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+				<Type>Time</Type>
+				<Operations>R</Operations>
+				<Mandatory>Optional</Mandatory>
+				<MultipleInstances>Single</MultipleInstances>
+				<Description><![CDATA[SUBSCRIBER ONLY UNIX Timestamp for when the last activity in the group was recorded, either by the node sending a message to the group our retrieved from the "Publishing Timestamp" of the last message received from the group. [default=0]]]></Description>
+				<Units/>
+			</Item>
+			<Item ID="2">
+				<Name>Most recent Sensor Notification</Name>
+				<Type>String</Type>
```
Comment (akeranen) [inline 3459.xml:64]: Also here, and other places with "links" [CoRE link](https://www.openmobilealliance.org/release/LightweightM2M/V1_2-20201110-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_2-20201110-A.html#11-0-Appendix-C-Data-Types-Normative) could make sense. But that would require later LwM2M version than 1.0 IIRC.

File: 3422.xml (hunk 1 above)
Comment (gmazzoleni) [reply 3422.xml]: Thank you Ari for pointing out.
You are perfectly right.
Changing to Integer (in order to keep compatibility with LWM2M 1.0).
Would it be fine?

File: 3452.xml (hunk 2 above)
Comment (gmazzoleni) [reply 3452.xml:98]: Thank you very much Ari and good point!
That could indeed make sense and could be considered very good practice.

Allow me to check internally to the Alliance if we can consider such change, in order to have a better formal description of the object, despite breaking compatibility with LWM2M 1.0

As an intermediate approach, could also be possible to keep String data type using Core Link syntax.

This also applies to some other resources as well, by the way (as you also commented below).

File: 3459.xml (hunk 4 above)
Comment (gmazzoleni) [reply 3459.xml:64]: Same as above. Thanks for pointing out

File: 3459.xml (hunk 3 above)
Comment (gmazzoleni) [reply 3459.xml:50]: Considering that data type is Floatign Point, we were assuming to use NaN floating point encoding.
In content formats such as TLV, JSON (and SenML-JSON) or CBOR (and SenML-CBOR) a standard representation exists.
I don't know about plain text syntax, though would assume "NaN"
Do you suggest any change in the approach?

File: 3452.xml (hunk 2 above)
Comment (gmazzoleni) [reply 3452.xml:98]: Plan is to keep String data type, but changing the description to mention the usage of Core Link format

**Comment by gmazzoleni** [comment]
Is there anything else needed from uCIFI side related to this pull request?
Otherwise, it could be merged.
Thank you.

**Pull request merged**